components/script/dom/textdecoderstream.rs

```diff
@@ -0,0 +1,190 @@
+use crate::dom::bindings::error::{Error, Fallible};
+use crate::dom::bindings::reflector::{Reflector, reflect_dom_object_with_proto};
+use crate::dom::bindings::root::{Dom, DomRoot};
+use crate::dom::bindings::str::DOMString;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::textdecodercommon::TextDecoderCommon;
+use crate::dom::transformstreamdefaultcontroller::TransformerType;
+use crate::dom::types::{TransformStream, TransformStreamDefaultController};
+use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
+
+/// <https://encoding.spec.whatwg.org/#decode-and-enqueue-a-chunk>
+#[allow(unsafe_code)]
+pub(crate) fn decode_and_enqueue_a_chunk(
+    cx: SafeJSContext,
+    global: &GlobalScope,
+    chunk: SafeHandleValue,
+    decoder: &TextDecoderCommon,
+    controller: &TransformStreamDefaultController,
+    can_gc: CanGc,
+) -> Fallible<()> {
+    // Step 1. Let bufferSource be the result of converting chunk to an AllowSharedBufferSource.
+    let conversion_result = unsafe {
+        ArrayBufferViewOrArrayBuffer::from_jsval(*cx, chunk, ()).map_err(|_| {
+            Error::Type("Unable to convert chunk into ArrayBuffer or ArrayBufferView".to_string())
+        })?
+    };
+    let buffer_source = conversion_result.get_success_value().ok_or(Error::Type(
+        "Unable to convert chunk into ArrayBuffer or ArrayBufferView".to_string(),
+    ))?;
+
+    // Step 2. Push a copy of bufferSource to decoder’s I/O queue.
+    // Step 3. Let output be the I/O queue of scalar values « end-of-queue ».
+    // Step 4. Implemented by `TextDecoderCommon::decode`, which uses the same procedure as
+    //      `TextDecoder`
+    let output_chunk = decoder.decode(Some(buffer_source), true)?;
+
+    // Step 4.2.2 If outputChunk is not the empty string, then enqueue
+    //      outputChunk in decoder’s transform.
+    if output_chunk.is_empty() {
+        return Ok(());
+    }
+    rooted!(in(*cx) let mut rval = UndefinedValue());
+    unsafe { output_chunk.to_jsval(*cx, rval.handle_mut()) };
+    controller.enqueue(cx, global, rval.handle(), can_gc)
+}
+
+/// <https://encoding.spec.whatwg.org/#flush-and-enqueue>
+#[allow(unsafe_code)]
+pub(crate) fn flush_and_enqueue(
+    cx: SafeJSContext,
+    global: &GlobalScope,
+    decoder: &TextDecoderCommon,
+    controller: &TransformStreamDefaultController,
+    can_gc: CanGc,
+) -> Fallible<()> {
+    // Step 1. Implemented by `TextDecoderCommon::decode` which uses a similar process
+    //      as `TextDecoder::Decode`.
+    let output_chunk = decoder.decode(None, false)?;
+
+    // Step 2.3.2 If outputChunk is not the empty string, then enqueue
+    //      outputChunk in decoder’s transform.
+    if output_chunk.is_empty() {
+        return Ok(());
+    }
+    rooted!(in(*cx) let mut rval = UndefinedValue());
+    unsafe { output_chunk.to_jsval(*cx, rval.handle_mut()) };
+    controller.enqueue(cx, global, rval.handle(), can_gc)
+}
```

Comment:
Missing "Otherwise, if result is [error](https://encoding.spec.whatwg.org/#error), [throw](https://webidl.spec.whatwg.org/#dfn-throw) a [TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror).".

Reply:
Perhaps, but the JS exception still needs to be set.